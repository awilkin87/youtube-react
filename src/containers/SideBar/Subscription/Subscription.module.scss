@use '../sidebar-common.modules';

.subscription {
  display: flex;
  justify-content: space-between;
  align-items: center;

  .leftContainer {
    display: flex;
    align-items: center;
    overflow: hidden;

    .label {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }

  // Icon class added by Semantic UI 
  :global(i.icon) {
    color: sidebar-common.$red;
    margin: 0;
  }

  .videoCount {
    color: sidebar-common.$grey;
  }
}