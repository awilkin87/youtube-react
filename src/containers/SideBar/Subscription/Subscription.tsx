import PropType, { InferProps } from 'prop-types';
import { Icon, Image } from 'semantic-ui-react';
import SideBarClickableItem from '../SideBarClickableItem/SideBarClickableItem';
import styles from './Subscription.module.scss';

/** React props for a sidebar Subscription. */
export type SidebarSubscriptionProps = InferProps<typeof propTypes>;

/** Clickable sidebar menu item for accessing and displaying information about a subscription. */
const Subscription: React.FunctionComponent<Readonly<SidebarSubscriptionProps>> = function (props) {
  let rightElement;
  if (props.broadcasting) {
    rightElement = <Icon name='podcast' />;
  } else {
    const newVideoCount = props.newVideos || 0;
    if (newVideoCount > 0) {
      rightElement = <span className={styles.videoCount}>{props.newVideos}</span>;
    }
  }

  return (
    <SideBarClickableItem>
      <div className={styles.subscription}>
        <span className={styles.leftContainer}>
          <Image avatar src='http://via.placeholder.com/28x28' />
          <span className={styles.label}>
            {props.name}
          </span>
        </span>
        {rightElement}
      </div>
    </SideBarClickableItem>
  );
}

const propTypes = {
  /** Subscription channel name. */
  name: PropType.string.isRequired,
  /** True if channel broadcasting live. */
  broadcasting: PropType.bool,
  /** Number of new videos not yet watched. */
  newVideos: PropType.number
}

Subscription.propTypes = propTypes;

export default Subscription;

/** Mock subscription while building out the app. */
export function MockSubscriptions() {
  return (
    <>
      <Subscription name='MusicChannel' broadcasting />
      <Subscription name='Coursera' newVideos={10} />
      <Subscription name='TEDx Talks' newVideos={23} />
      <Subscription name='Stanford iOS' />
      <Subscription name='Udacity' newVideos={114} />
      <Subscription name='Best Most Awesomest Channel Ever' newVideos={5} />
    </>
  );
}
