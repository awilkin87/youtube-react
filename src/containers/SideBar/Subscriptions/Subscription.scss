@import '../../../styles/shared.scss';

.subscription {
  display: flex;
  justify-content: space-between;
  align-items: center;

  .subscription-left-container {
    display: flex;
    align-items: center;
    overflow: hidden;

    .subscription-name {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }

  i.icon {
    color: $red;
    margin: 0;
  }

  .new-video-count {
    color: $grey;
  }
}