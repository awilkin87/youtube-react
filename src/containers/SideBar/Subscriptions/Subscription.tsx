import { Icon, Menu, Image } from 'semantic-ui-react';
import PropType, { InferProps } from 'prop-types';
import './Subscription.scss';

export type SidebarSubscriptionProps = InferProps<typeof Subscription.propTypes>;

function Subscription(props: Readonly<SidebarSubscriptionProps>) {
  let rightElement;
  if (props.broadcasting) {
    rightElement = <Icon name='podcast' />;
  } else {
    const newVideoCount = props.newVideos || 0;
    if (newVideoCount > 0) {
      rightElement = <span className='new-video-count'>{props.newVideos}</span>;
    }
  }

  return (
    <Menu.Item className='sidebar-clickable'>
      <div className='subscription'>
        <span className='subscription-left-container'>
          <Image avatar src='http://via.placeholder.com/28x28' />
          <span className='subscription-name'>
            {props.name}
          </span>
        </span>
        {rightElement}
      </div>
    </Menu.Item>
  );
}

Subscription.propTypes = {
  name: PropType.string.isRequired,
  broadcasting: PropType.bool,
  newVideos: PropType.number
}

export function MockSubscriptions() {
  return (
    <>
      <Subscription name='MusicChannel' broadcasting />
      <Subscription name='Coursera' newVideos={10} />
      <Subscription name='TEDx Talks' newVideos={23} />
      <Subscription name='Stanford iOS' />
      <Subscription name='Udacity' newVideos={114} />
      <Subscription name='Best Most Awesomest Channel Ever' newVideos={5} />
    </>
  );
}

export default Subscription;