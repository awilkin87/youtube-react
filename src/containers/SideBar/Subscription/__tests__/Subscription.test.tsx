import { shallow } from "enzyme";
import Subscription from '../Subscription';

describe('SideBar Subscription', () => {
  test('Renders Subscription w/ no videos and not broadcasting', () => {
    const wrapper = shallow(<Subscription name="MySubscription" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Renders Subscription w/ new videos and not broadcasting', () => {
    const wrapper = shallow(<Subscription name="MySubscription" newVideos={10} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Renders Subscription w/ no videos and is broadcasting', () => {
    const wrapper = shallow(<Subscription name="MySubscription" broadcasting />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Renders Subscription w/ new videos and is broadcasting', () => {
    const wrapper = shallow(<Subscription name="MySubscription" newVideos={10} broadcasting />);
    expect(wrapper).toMatchSnapshot();
  });
});