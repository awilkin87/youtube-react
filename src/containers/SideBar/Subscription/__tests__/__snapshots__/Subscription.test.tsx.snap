// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar Subscription Renders Subscription w/ new videos and is broadcasting 1`] = `
<SideBarClickableItem>
  <div
    className="subscription"
  >
    <span
      className="leftContainer"
    >
      <Image
        as="img"
        avatar={true}
        src="http://via.placeholder.com/28x28"
        ui={true}
      />
      <span
        className="label"
      >
        MySubscription
      </span>
    </span>
    <Icon
      as="i"
      name="podcast"
    />
  </div>
</SideBarClickableItem>
`;

exports[`SideBar Subscription Renders Subscription w/ new videos and not broadcasting 1`] = `
<SideBarClickableItem>
  <div
    className="subscription"
  >
    <span
      className="leftContainer"
    >
      <Image
        as="img"
        avatar={true}
        src="http://via.placeholder.com/28x28"
        ui={true}
      />
      <span
        className="label"
      >
        MySubscription
      </span>
    </span>
    <span
      className="videoCount"
    >
      10
    </span>
  </div>
</SideBarClickableItem>
`;

exports[`SideBar Subscription Renders Subscription w/ no videos and is broadcasting 1`] = `
<SideBarClickableItem>
  <div
    className="subscription"
  >
    <span
      className="leftContainer"
    >
      <Image
        as="img"
        avatar={true}
        src="http://via.placeholder.com/28x28"
        ui={true}
      />
      <span
        className="label"
      >
        MySubscription
      </span>
    </span>
    <Icon
      as="i"
      name="podcast"
    />
  </div>
</SideBarClickableItem>
`;

exports[`SideBar Subscription Renders Subscription w/ no videos and not broadcasting 1`] = `
<SideBarClickableItem>
  <div
    className="subscription"
  >
    <span
      className="leftContainer"
    >
      <Image
        as="img"
        avatar={true}
        src="http://via.placeholder.com/28x28"
        ui={true}
      />
      <span
        className="label"
      >
        MySubscription
      </span>
    </span>
  </div>
</SideBarClickableItem>
`;
